tests/beta/interfaces/test_tape_tf.py

```diff
@@ -785,3 +783,571 @@ def test_sampling(self):
+        def circuit(a):
+            qml.RY(a[0], wires=0)
+            qml.RX(a[1], wires=0)
+            return expval(qml.PauliZ(0))
+
+        with tf.GradientTape() as tape:
+            res = circuit(a)
+
+        tape.jacobian(res, a)
+
+        for args in spy.call_args_list:
+            assert args[1]["order"] == 2
+            assert args[1]["h"] == 1e-8
+
+    def test_changing_trainability(self, dev_name, diff_method, mocker, tol):
+        """Test changing the trainability of parameters changes the
+        number of differentiation requests made"""
+        if diff_method == "backprop":
+            pytest.skip("Test does not support backprop")
+
+        a = tf.Variable(0.1, dtype=tf.float64)
+        b = tf.Variable(0.2, dtype=tf.float64)
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qnode(dev, interface="tf", diff_method="finite-diff")
+        def circuit(a, b):
+            qml.RY(a, wires=0)
+            qml.RX(b, wires=1)
+            qml.CNOT(wires=[0, 1])
+            return expval(qml.PauliZ(0)), expval(qml.PauliY(1))
+
+        with tf.GradientTape() as tape:
+            res = circuit(a, b)
+
+        # the tape has reported both gate arguments as trainable
+        assert circuit.qtape.trainable_params == {0, 1}
+
+        expected = [tf.cos(a), -tf.cos(a) * tf.sin(b)]
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        spy = mocker.spy(QuantumTape, "numeric_pd")
+
+        jac = tape.jacobian(res, [a, b])
+        expected = [
+            [-tf.sin(a), tf.sin(a) * tf.sin(b)],
+            [0, -tf.cos(a) * tf.cos(b)],
+        ]
+        assert np.allclose(jac, expected, atol=tol, rtol=0)
+
+        # QuantumTape.numeric_pd has been called for each argument
+        assert len(spy.call_args_list) == 2
+
+        # make the second QNode argument a constant
+        a = tf.Variable(0.54, dtype=tf.float64)
+        b = tf.constant(0.8, dtype=tf.float64)
+
+        with tf.GradientTape() as tape:
+            res = circuit(a, b)
+
+        # the tape has reported only the first argument as trainable
+        assert circuit.qtape.trainable_params == {0}
+
+        expected = [tf.cos(a), -tf.cos(a) * tf.sin(b)]
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        spy.call_args_list = []
+        jac = tape.jacobian(res, a)
+        expected = [-tf.sin(a), tf.sin(a) * tf.sin(b)]
+        assert np.allclose(jac, expected, atol=tol, rtol=0)
+
+        # QuantumTape.numeric_pd has been called only once
+        assert len(spy.call_args_list) == 1
+
+    def test_classical_processing(self, dev_name, diff_method, tol):
+        """Test classical processing within the quantum tape"""
+        a = tf.Variable(0.1, dtype=tf.float64)
+        b = tf.constant(0.2, dtype=tf.float64)
+        c = tf.Variable(0.3, dtype=tf.float64)
+
+        dev = qml.device(dev_name, wires=1)
+
+        @qnode(dev, diff_method=diff_method, interface="tf")
+        def circuit(a, b, c):

```

Comment:
Very minor detail, but naming these the same as the `tf.Variable` objects could potentially shadow those names from the outer scope. Maybe change to `x`, `y` and `z` for the qfunc?